design/sticky-session-design.md

```diff
@@ -0,0 +1,170 @@
+# Sticky session support
+
+**Status**: _draft_
+
+This document proposes adding support for sticky session in contour.
+
+## Goals
+
+- Enabling support for sticky session in Contour using
+    - Cookie
+    - Source IP
+    - Header
+- Adding support in IngressRoute CRD. 
+
+## Non-goals
+
+- Making options in IngressRoute very specific to Envoy is not a goal. Options in IngressRoute will be generic.
+
+## Background
+
+Sticky-sessions allow requests to be routed to the same host based on certain parameters of the request.
+The parameters can be a cookie, the source IP, a custom header, etc.
+
+Envoy supports sticky sessions by using a hash-based load balancing strategy (RingHash, Maglev). 
+Specifying a *hash policy* for such load-balancers will enable sticky sessions. However, we are not aiming at making
+this very specific to Envoy. IngressRoute options should be generic enough without exposing
+internals of Envoy itself.
+
+
+## High-Level Design
+
+Sticky sessions are enabled per route. The following change are required in IngressRoute:
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata: 
+  name: basic
+  namespace: default
+spec: 
+  virtualhost:
+    fqdn: foo-basic.bar.com
+  routes: 
+    - match: /
+      sticky-session:
```

Comment:
We already have the 'strategy:' key defined, why not add "sticky-session" as a new option?

Also, a route can have multiple services, which have a weighting -- not a load balancing strategy. Is it possible to apply stick sessions across multiple clusters -- i don't think it is so it won't be possible to apply sticky sessions across a weighted set of clusters.

Reply:
We can use the 'strategy' key to specify 'sticky-session' as an option. However since strategy is of string type, we'd need another key to specify options for this strategy. Do you suggest we have another key: *strategy-options* (or any other key perhaps) under which we could specify the option to use: cookie/ip/header?